File: bootstrap/src/main/java/net/dontdrinkandroot/wicket/bootstrap/component/navbar/RepeatingNavbarNav.java
/*
 * Copyright (C) 2012-2017 Philip Washington Sorst <philip@sorst.net>
 * and individual contributors as indicated
 * by the @authors tag.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dontdrinkandroot.wicket.bootstrap.component.navbar;

import net.dontdrinkandroot.wicket.behavior.CssClassAppender;
import net.dontdrinkandroot.wicket.bootstrap.component.nav.AbstractRepeatingNav;
import net.dontdrinkandroot.wicket.bootstrap.css.BootstrapCssClass;
import org.apache.wicket.markup.ComponentTag;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class RepeatingNavbarNav<T> extends AbstractRepeatingNav<T>
{
    public RepeatingNavbarNav(String id)
    {
        super(id);
    }

    @Override
    protected void onInitialize()
    {
        super.onInitialize();

        this.add(new CssClassAppender(BootstrapCssClass.NAVBAR_NAV));
    }

    @Override
    protected void onComponentTag(ComponentTag tag)
    {
        tag.setName("ul");
        super.onComponentTag(tag);
    }
}

File: bootstrap/src/main/kotlin/net/dontdrinkandroot/wicket/bootstrap/component/navbar/RepeatingNavbarNav.kt
package net.dontdrinkandroot.wicket.bootstrap.component.navbar

import net.dontdrinkandroot.wicket.behavior.CssClassAppender
import net.dontdrinkandroot.wicket.bootstrap.component.item.ItemView
import net.dontdrinkandroot.wicket.bootstrap.component.nav.AbstractRepeatingNav
import net.dontdrinkandroot.wicket.bootstrap.css.BootstrapCssClass
import org.apache.wicket.behavior.Behavior
import org.apache.wicket.markup.ComponentTag
import org.apache.wicket.model.IModel

abstract class RepeatingNavbarNav<T>(
    id: String,
    model: IModel<T>? = null,
    vararg behaviors: Behavior
) : AbstractRepeatingNav<T>(id, model, *behaviors) {

    init {
        add(CssClassAppender(BootstrapCssClass.NAVBAR_NAV))
        add(*behaviors)
    }

    override fun onComponentTag(tag: ComponentTag) {
        tag.name = "ul"
        super.onComponentTag(tag)
    }
}

fun <T> repeatingNavbarNav(
    id: String,
    model: IModel<T>? = null,
    vararg behaviors: Behavior,
    populateItemsHandler: ItemView.() -> Any?
) = object : RepeatingNavbarNav<T>(id, model, *behaviors) {
    override fun populateItems(itemView: ItemView) {
        populateItemsHandler(itemView)
    }
}

fun repeatingNavbarNav(
    id: String,
    vararg behaviors: Behavior,
    populateItemsHandler: ItemView.() -> Any?
) = object : RepeatingNavbarNav<Void>(id, null, *behaviors) {
    override fun populateItems(itemView: ItemView) {
        populateItemsHandler(itemView)
    }
}

fun NavbarCollapseView.navbarNav(
    vararg behaviors: Behavior,
    populateItemsHandler: ItemView.() -> Any?
) {
    add(object : RepeatingNavbarNav<Void>(newChildId(), null, *behaviors) {
        override fun populateItems(itemView: ItemView) {
            populateItemsHandler(itemView)
        }
    })
}